modules/FvwmBanner/FvwmBanner.man
.\" t
.\" @(#)FvwmBanner.1	1/12/94
.TH FvwmBanner 1.20 "Jan 28, 1994"
.UC
.SH NAME
FvwmBanner \- the FVWM Banner
.SH SYNOPSIS
FvwmBanner is intended to be spawned by fvwm.

.SH DESCRIPTION
FvwmBanner displays an Fvwm Logo in the center of the screen
for 3 seconds.

.SH COPYRIGHTS
None.

.SH INITIALIZATION
Nothing interesting.

.SH INVOCATION
FvwmBanner can be invoked by binding the action 'Module FvwmBanner' to
a menu or key-stroke in the .fvwmrc file.  Fvwm will search directory
specified in the ModulePath configuration option to attempt to locate
FvwmBanner. Although nothing keeps you from launching FvwmBanner at
start-up time, you probably don't want to.  You can also give it an
optional file parameter, like 'FvwmBanner doomface.xpm' or spcify an
alternate default pixmap via configuration options.

.SH CONFIGURATION OPTIONS

.IP "*FvwmBannerPixmap \fIfile\fP"
Tells the module to display \fIfile\fP instead of the built in pixmap.

.IP "*FvwmBannerTimeout \fIsec\fP"
Tells the module to display for \fIsec\fP seconds instead of default of 3.


.SH AUTHOR
Robert Nation


ChangeLog
2026-02-23 David Flater

	Renamed package to Fvwm95 Forever and moved to GitHub.
	PACKAGE_VERSION = "Forever"

	Fixed FTBs for GCC 15.
	Updated things in configure.ac, Makefile.am, and system.fvwm95rc.in.
	Converted README to Markdown and revised as necessary.
	Merged BUGS and NEVERDO into README.md.

2020-12-10 David Flater

	fvwm95-2.0.43f-dwf8

	* Fixed multiple definitions of global variables in modules
	FvwmScroll and FvwmTaskBar.  GCC 10 changed the default behavior
	from -fcommon to -fno-common, which caused these multiple
	definitions to be rejected.

2020-03-18 David Flater

	fvwm95-2.0.43f-dwf7

	* Fixed buffer overflows in fvwm/read.c SendDataToModule which
	were triggered by a long IconPath or PixmapPath.

	* Added mini-firefox.xpm.

	* Updated system.fvwm95rc.in:
	- Added Firefox;
	- Added dosbox-x to the workaround for SDL2 fullscreen lossage;
	- Disabled Alt-arrow hotkeys (shortcuts to scroll by 1/10 of a
	page) to stop conflict with Doom (Alt = strafe).

	* Updated INSTALL.

2018-01-22 David Flater

	fvwm95-2.0.43f-dwf6

	* Updated system.fvwm95rc.in:
	- Fixed usage of killall;
	- Added the fullscreen workaround to chocolate-doom etc.;
	- Comments.

	* Updated INSTALL.

2017-04-20 David Flater

	fvwm95-2.0.43f-dwf5

	* Replaced faulty recursive build system with monolithic Automake:
        - Added script utils/makemakefileam.sh to generate Makefile.am;
	- Replaced configure.ac;
	- Purged all Imakefile and Makefile.in;
	- Adjusted install and search paths to agree with modern Linux.

	* Fixed innumerable instances of overflow, underflow, corruption,
	undefined behavior, division by zero, and broken parsing which
	were responsible for crashes and misbehavior in fvwm95 and every
	module.

	* Cleaned up compiler warnings through -Wall -Wno-parentheses on
	GCC 6.3.0.  Left the following:
	  ButtonArray.c:
	    warning: statement with no effect (BOGUS)
          etc:
  	    warning: the use of `mktemp' is dangerous, better use `mkstemp'
	    warning: implicit declaration of function 'ConsoleMessage' (BOGUS)

	* Harmonized swallow triggers in FvwmButtons, FvwmWharf,
	FvwmTaskBar, and FvwmScript (SwallowExec) to match on resource
	name (Resource), resource class (Class), or window name (Name),
	but not icon name (Icon Name).

	* Fixed fvwm95 bugs:
	- windows suddenly becoming enormous when resizing occurs near the
	  edge of the virtual desktop;
	- full-screen glitch when resizing a window that was declared with
  	  NoHandles and BorderWidth 0;
	- UseStyle option not accepting quoted identifiers;
	- false matching by UseStyle option;
	- no default WindowManagerName for Restart command (now fvwm95).

	* Fixed FvwmAudio bugs:
	- playing wrong or no sounds;
	- hanging, waiting on subprocesses to exit;
	- breaking audio_filename at whitespace.

	* Fixed FvwmConsole not using libreadline by default.

	* Fixed FvwmScript and sample scripts bugs:
	- missing default for FvwmScriptPath;
	- references to non-existent things in Buttons script;
	- infinite loop in PointerSetup script (but it is hopelessly fragile
	  and may hang on another system).

	* Fixed FvwmTaskBar bugs:
	- swallowing completely broken;
	- animation broken or too fast;
	- main event loop complexity;
	- no warning on missing *FvwmTaskBarGoodyLoadableSymbol;
	- bad geometry, location, and spacing of goodies;
	- bad calibration of x coordinates on mouse clicks on goodies;
	- area cleared different than area drawn;
	- missing default for FvwmTaskBarModuleIconPathBatStat;
	- only allowing one directory in FvwmTaskBarModuleIconPathBatStat.

	* Fixed FvwmWharf bugs:
	- black box over first button when swallowing failed;
	- wrong background colors on 32-bit arch;
	- animation too fast.

	* Corrected inaccuracies in man pages and HTML docs:
	- Documented the swallow commands for FvwmTaskBar;
	- Corrected description of the swallow commands for FvwmWharf;
	- Fixed ghost reference to FvwmInitBanner;
	- Corrected description of the name-matching behavior of the Style
  	  command to agree with what is actually done in fvwm/add_window.c
	  LookInList.

	* Made a few minor additions and enhancements:
	- Replaced the sample/default system.fvwm95rc;
	- Made FvwmTaskBar goodies work by single-click;
	- Improved error reporting by GoodyLoadable.c;
	- Added mini-xemacs.xpm, mini-chrome.xpm, mini-chromium.xpm,
	  mini-wilber16.xpm, and green.xpm;
	- Added LogoTest script.

	* Cleaned up random things:
	- Merged all needed preprocessor defines into FVWMconfig.h and made
	  all sources include it;
	- Removed version.h headers and declarations that conflicted with
	  FVWMconfig.h;
	- Merged index.html and GPL.html in with the other HTML files;
	- Merged xpmroot into utils;
	- Merged and purged duplicative icon collections:
	  - FvwmScript/Scripts/some_nice_icons:
	      moved shutdown.xpm to pixmaps/shutdown2.xpm
	      moved window.xpm to pixmaps/window2.xpm
	  - FvwmTaskBar/icons:
	      moved mini-mail.xpm to mini-icons/mini-mail2.xpm
	      moved minimail.xbm to mini-icons/minimail.xbm
	- Fixed saferealloc and made UpdateString use it (3 copies);
	- Moved random READMEs, example files, etc. from module
	  subdirectories into docs/modules;
	- Moved other random READMEs, example files, etc. into docs/random;
	- Moved the untested/unmaintained ASSound and OffiX subdirs of
	  FvwmWharf to docs/modules/FvwmWharf;
	- Deleted all other unused source files and scripts;
	- Normalized file permissions;
	- Stripped trailing whitespace.

	Not done:  see BUGS, NEVERDO.

2014-01-26 David Flater

	fvwm95-2.0.43f-dwf4

	* Fixed BatStatModule showing "charging" icon instead of "online"
	when fully charged.

	* Changed the "alert" and "empty" levels from 10% and 5% to 7% and
	2% respectively.

2014-01-25 David Flater

 	fvwm95-2.0.43f-dwf3

	* Fixed crash of FvwmTaskBar when a goody (BatStatModule) returned
	NULL from its init function to signify failure to init.

	* Updated BatStatModule to work with /sys/class/power_supply.

modules/FvwmBanner/FvwmBanner.c
#include <FVWMconfig.h>
/***************************************************************************
 * FvwmBanner
 *
 *  Show Fvwm Banner
 *
 ***************************************************************************/

#ifdef ISC
#include <sys/bsdtypes.h> /* Saul */
#endif

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif

#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>

#include <X11/StringDefs.h>
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#include <X11/extensions/shape.h>

#include <X11/xpm.h>


#ifdef BROKEN_SUN_HEADERS
#include "../../fvwm/sun_headers.h"
#endif

#include <fvwm/fvwmlib.h>

#ifdef NEEDS_ALPHA_HEADER
#include "../../fvwm/alpha_header.h"
#endif /* NEEDS_ALPHA_HEADER */

#include "../../pixmaps/fvwm95_big.xpm"
#if 0
#include "../../pixmaps/k2.xpm"
#endif /* 0 */

typedef struct _XpmIcon {
    Pixmap pixmap;
    Pixmap mask;
    XpmAttributes attributes;
}        XpmIcon;

/**************************************************************************
 * A few function prototypes
 **************************************************************************/
void RedrawWindow(void);
void GetXPMData(char **);
void GetXPMFile(char *,char *);
void change_window_name(char *str);
int flush_expose (Window w);
static void parseOptions (int fd[2]);

XpmIcon view;
Window win;

char *pixmapPath = NULL;
char *pixmapName = NULL;
char *myName = NULL;

int timeout = 3000000; /* default time of 3 seconds */

Display *dpy;			/* which display are we talking to */
Window Root;
int screen;
int x_fd;
int d_depth;
int ScreenWidth, ScreenHeight;
XSizeHints mysizehints;
Pixel back_pix, fore_pix;
GC NormalGC,FGC;
static Atom wm_del_win;
Colormap colormap;

#define MW_EVENTS   (ExposureMask | ButtonReleaseMask)

/****************************************************************************
 *
 * Creates an icon window as needed
 *
 ****************************************************************************/
int main(int argc, char **argv)
{
  char *display_name = NULL, *string = NULL;
  int retval=0;
  XEvent Event;
  fd_set in_fdset;
  int fd_width ;
  struct timeval value;
  int fd[2];

  fd_width = GetFdWidth();

  /* Save our program  name - for error messages */
  string = strrchr (argv[0], '/');
  if (string != (char *) 0) string++;

  myName = safemalloc (strlen (string) + 1);
  strcpy (myName, string);

  if(argc>=3)
  {
      /* sever our connection with fvwm, if we have one. */
      fd[0] = atoi(argv[1]);
      fd[1] = atoi(argv[2]);

#if 0
      if(fd[0]>0)close(fd[0]);
      if(fd[1]>0)close(fd[1]);
#endif /* 0 */
  }
  else
  {
    fprintf (stderr,
	     "%s version %s should only be executed by fvwm!\n",
	     myName,
	     VERSION);
    exit(1);
  }

  if (argc > 6) {
    pixmapName = safemalloc (strlen (argv[6]) + 1);
    strcpy (pixmapName, argv[6]);
  }

  /* Open the display */
  if (!(dpy = XOpenDisplay(display_name)))
    {
      fprintf(stderr,"FvwmBanner: can't open display %s",
	      XDisplayName(display_name));
      exit (1);
    }
  screen= DefaultScreen(dpy);
  Root = RootWindow(dpy, screen);
  colormap = XDefaultColormap(dpy,screen);
  d_depth = DefaultDepth(dpy, screen);
  x_fd = XConnectionNumber(dpy);

  ScreenHeight = DisplayHeight(dpy,screen);
  ScreenWidth = DisplayWidth(dpy,screen);

  parseOptions(fd);

  /* Get the xpm banner */
  if (pixmapName)
    GetXPMFile(pixmapName,pixmapPath);
  else
#if 0
    if(d_depth > 4)
      GetXPMData(k2_xpm);
    else
#endif /* 0 */
      GetXPMData(fvwm95_big_xpm);

  /* Create a window to hold the banner */
  mysizehints.flags=
    USSize|USPosition|PWinGravity|PResizeInc|PBaseSize|PMinSize|PMaxSize;
  /* subtract one for the right/bottom border */
  mysizehints.width = view.attributes.width;
  mysizehints.height=view.attributes.height;
  mysizehints.width_inc = 1;
  mysizehints.height_inc = 1;
  mysizehints.base_height = mysizehints.height;
  mysizehints.base_width = mysizehints.width;
  mysizehints.min_height = mysizehints.height;
  mysizehints.min_width = mysizehints.width;
  mysizehints.max_height = mysizehints.height;
  mysizehints.max_width = mysizehints.width;
  mysizehints.win_gravity = NorthWestGravity;

  mysizehints.x = (ScreenWidth - view.attributes.width)/2;
  mysizehints.y = (ScreenHeight - view.attributes.height)/2;

  win = XCreateSimpleWindow(dpy,Root,mysizehints.x,mysizehints.y,
				 mysizehints.width,mysizehints.height,
				 1,fore_pix ,None);


  /* Set assorted info for the window */
  XSetTransientForHint(dpy,win,Root);
  wm_del_win = XInternAtom(dpy,"WM_DELETE_WINDOW",False);
  XSetWMProtocols(dpy,win,&wm_del_win,1);

  XSetWMNormalHints(dpy,win,&mysizehints);
  change_window_name("FvwmBanner");

  XSetWindowBackgroundPixmap(dpy,win,view.pixmap);
#ifdef SHAPE
  if(view.mask != None)
    XShapeCombineMask(dpy, win, ShapeBounding,0,0,view.mask, ShapeSet);
#endif
  XMapWindow(dpy,win);
  XSync(dpy,0);
#if 0
  sleep_a_little(timeout);
#else
  XSelectInput(dpy,win,ButtonReleaseMask);
  /* Display the window */
  value.tv_usec = timeout % 1000000;
  value.tv_sec = timeout / 1000000;
  while(1)
  {
    FD_ZERO(&in_fdset);
    FD_SET(x_fd,&in_fdset);

    if(!XPending(dpy))
#ifdef __hpux
      retval=select(fd_width,(int *)&in_fdset, 0, 0, &value);
#else
      retval=select(fd_width,&in_fdset, 0, 0, &value);
#endif
    if (retval==0)
    {
      XDestroyWindow(dpy,win);
      XSync(dpy,0);
      exit(0);
    }

    if(FD_ISSET(x_fd, &in_fdset))
    {
      /* read a packet */
      XNextEvent(dpy,&Event);
      switch(Event.type)
      {
        case ButtonRelease:
          XDestroyWindow(dpy,win);
          XSync(dpy,0);
          exit(0);
        case ClientMessage:
          if (Event.xclient.format==32 && Event.xclient.data.l[0]==wm_del_win)
          {
            XDestroyWindow(dpy,win);
            XSync(dpy,0);
            exit(0);
          }
        default:
          break;
      }
    }
  }
#endif /* 0 */
  return 0;
}

/****************************************************************************
 *
 * Looks for a color XPM icon file
 *
 ****************************************************************************/
void GetXPMData(char **data)
{
  view.attributes.valuemask = XpmReturnPixels| XpmCloseness | XpmExtensions;
  view.attributes.closeness = 40000 /* Allow for "similar" colors */;
  if(XpmCreatePixmapFromData(dpy, Root, data,
                             &view.pixmap, &view.mask,
                             &view.attributes)!=XpmSuccess)
  {
    fprintf(stderr,"FvwmBanner: ERROR couldn't convert data to pixmap\n");
    exit(1);
  }
}
void GetXPMFile(char *file, char *path)
{
  char *full_file=NULL;

  view.attributes.valuemask = XpmReturnPixels| XpmCloseness | XpmExtensions;
  view.attributes.closeness = 40000 /* Allow for "similar" colors */;

  if (file)
    full_file = findIconFile(file,path,R_OK);

  if (full_file)
  {
    if(XpmReadFileToPixmap(dpy,
                           Root,
                           full_file,
                           &view.pixmap,
                           &view.mask,
                           &view.attributes) == XpmSuccess)
    {
      return;
    }
    fprintf(stderr,"FvwmBanner: ERROR reading pixmap file\n");
  }
  else
    fprintf(stderr,"FvwmBanner: ERROR finding pixmap file in PixmapPath\n");
  GetXPMData(fvwm95_big_xpm);
}

void nocolor(char *a, char *b)
{
 fprintf(stderr,"FvwmBanner: can't %s %s\n", a,b);
}

static void parseOptions (int fd[2])
{
  char *tline= NULL;
  int   clength;

  clength = strlen (myName);

  while (GetConfigLine (fd, &tline),tline != NULL)
  {
    if (strlen (tline) > 1)
    {
      if (strncasecmp (tline,
			 CatString3 ("*", myName, "Pixmap"),
			 clength + 7) ==0)
      {
	if (pixmapName == (char *) 0)
        {
	  CopyString (&pixmapName, &tline[clength+7]);
	  if (pixmapName[0] == 0)
          {
	    free (pixmapName);
	    pixmapName = (char *) 0;
	  }
	}
        continue;
      }
      if (strncasecmp (tline,
			 CatString3 ("*", myName, "Timeout"),
			 clength + 8) ==0)
      {
        timeout = atoi(&tline[clength+8]) * 1000000;
        continue;
      }
      if (strncasecmp(tline, "PixmapPath",10)==0)
      {
        CopyString (&pixmapPath, &tline[10]);
        if (pixmapPath[0] == 0)
        {
          free (pixmapPath);
          pixmapPath = (char *) 0;
        }
        continue;
      }
    }
  }
  return;
}

/**************************************************************************
 *  Change the window name displayed in the title bar.
 **************************************************************************/
void change_window_name(char *str)
{
  XTextProperty name;

  if (XStringListToTextProperty(&str,1,&name) == 0)
    {
      fprintf(stderr,"FvwmBanner: cannot allocate window name");
      return;
    }
  XSetWMName(dpy,win,&name);
  XSetWMIconName(dpy,win,&name);
  XFree(name.value);
}


/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 ***********************************************************************/

/*ARGSUSED*/
void DeadPipe (int nonsense)
{
  exit (0);
}
